### DISCUSSION TITLE
Using AAR dependencies

### DISCUSSION THREAD

**Discussion by llh-him** [shreyashsaitwal/rush-cli#37] [Q&A]
Hi, I am working with an extension that needs to use external library. As I have read through the wiki page, which states that JAR and AAR could be placed inside the deps folder. However, the AAR file could not be used, only the JAR does.

**Comment by shreyashsaitwal** [comment]
That's correct, you can't use AAR libraries in Rush; only JARs. This is because App Inventor doesn't yet allow extensions to use AAR libraries.

However, you can unzip the required AAR library and use the `classes.jar` JAR file from it as your extension's dependency. You might also want to merge the Android manifest of the AAR with that of your extension.

But keep in mind that this workaround only works for AARs that don't depend on any of the resource files bundled within them.

**Reply by llh-him** [reply]
Thanks for your reply. Well, it seems that I need to get some coffee to get it works then.

May I ask one more question as I am new to app inventor extension development. What would be the solution if some of the code needs to run with @AfterView annotation in native code? As I found that the code wouldn't run if I placed in constructor.

**Reply by shreyashsaitwal** [reply]
You might want to create a `Initialize` method block and put all of your view initialization logic there. Users of your extension would then call this block from the `Screen.Initialize` event block.

But generally speaking, most of the time you don't need this. Placing the initialization logic in the constructor body should work. Can you show me the code that you're having trouble with placing in the constructor body?

**Reply by llh-him** [reply]
I have tried in this way, but it is not loading as expected. Maybe I need to have a look on the library to see how it is called
![Screenshot 2022-05-19 16 11 30](https://user-images.githubusercontent.com/67906816/169245550-ddd09555-76ba-4bfb-ad55-c3a39bdf321a.png)
![Screenshot 2022-05-19 16 12 32](https://user-images.githubusercontent.com/67906816/169245741-376bad94-2ac9-4350-bc94-e0abc7047a64.png)
The Listener is referenced to the library.

**Reply by shreyashsaitwal** [reply]
> I have tried in this way, but it is not loading as expected.

Do you get any runtime errors? To me, everything seems fine.

**Reply by llh-him** [reply]
No. I have tried adding log to logcat also, there is no any records. Not sure the native code of logging works with ai companion or not, but just nothing logged.

The `WhenLog()` event won't despatch also when the code is called in constructor or called the `init()` through `screen.initialize` block. The most weird thing is if I called `init()` in a click event of button, it does call `WhenLog()` at least.

**Reply by shreyashsaitwal** [reply]
Logging to the logcat should work. Can you show me how you're trying to do it? And where does the `GestureView` gets created on the screen? You don't seem to be giving it a parent layout/component to place itself in.

**Reply by llh-him** [reply]
The code below is used for logging in logcat.
`private void Log(String msg) { Log.i(TAG, msg);}` where TAG is the package name.

GestureView got three kinds of constructor,
`public GestureView(android.content.Context context)`
`public GestureView(android.content.Context context, android.util.AttributeSet attrs)`
`public GestureView(android.content.Context context, android.util.AttributeSet attrs, int defStyleAttr)`

Thus, I passed the context into it by using the below code inside constructor
`activity = container.$context();`

The GestureView is initialized in `init()` method

**Reply by llh-him** [reply]
As it is using GestureView which is using `import android.view.GestureDetector;`, would it be considered the extension using Visible component instead of AndroidNonvisibleComponent?

**Reply by shreyashsaitwal** [reply]
No, only the components visible in the designer section of AI2 when you drag and drop them are considered visible, everything else is an `AndroidNonvisibleComponent`.

Can you give me a link to the library to which `GestureView` belongs? It's tough to give any advice unless I know for sure how exactly the said class works.

Also, you might want to check out this article where I guide you through how to manipulate a view component. It is old and therefore not using Rush, but except for a few parts, it's completely relevant even when using Rush.

Article: https://saitwalshreyash19.medium.com/writing-your-first-app-inventor-2-extension-dc6d5d4ff824

**Reply by llh-him** [reply]
https://sdk.madgaze.com/x5/android-sdk Here is the sdk used for my development.

I have look through the android api about gesture detector. It seems that the gesture detector needs the context of activity or view.
